src/client/components/stage-details/stage-details.component.js

```diff
@@ -0,0 +1,173 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import './stage-details.style.css';
+import SectionHeader from '../section-header/section-header.component';
+import Button from '../button/button.component';
+
+export default function StageDetails({
+  cropId,
+  stages,
+  levels,
+  data,
+  onChange,
+}) {
+  const PARAMETERS = {
+    temperature: 'Temperature (°C)',
+    humidity: 'Humidity (g/m3)',
+    ph: 'PH',
+    ec: 'EC (ppm)',
+    // eslint-disable-next-line @typescript-eslint/camelcase
+    water_level: 'Water level (cm)',
+  };
+
+  // Get the parameter data for crop with id of cropId
+  const [cropData, setCropData] = useState([]);
+
+  useEffect(() => {
+    setCropData(
+      data.filter((dataPoint) => parseInt(dataPoint.fk_crop_id, 10) === cropId),
+    );
+  }, [cropId, data]);
+
+  // Select the last stage to display as default
+  const lastStage = Math.max(...stages.map((s) => s.id));
+  const [selectedStageIndex, setSelectedStageIndex] = useState(lastStage);
+
+  // Get data for the currently selected stage
+  const [dataForStage, setDataForStage] = useState([]);
+
+  useEffect(() => {
+    const selectedStages = cropData.filter(
+      (dataPoint) =>
+        parseInt(dataPoint.fk_crop_stage_id, 10) === selectedStageIndex,
+    );
+    setDataForStage(selectedStages);
+  }, [selectedStageIndex, cropData]);
+
+  function handleInputChange(e, level, parameter) {
+    const valueEntered = e.target.value;
+    e.preventDefault();
+    setCropData((prev) => {
+      return updateCropDataOnInput(prev, valueEntered, level, parameter);
+    });
+  }
+
+  function updateCropDataOnInput(prevData, valueEntered, level, parameter) {
+    return prevData.map((dataPoint) => {
+      if (
+        Number(dataPoint.fk_crop_stage_id) === selectedStageIndex &&
+        dataPoint.parameter === parameter
+      ) {
+        const updatedDataPoint = {
+          ...dataPoint,
+          [`${level}_value`]: valueEntered,
+        };
+        return updatedDataPoint;
+      }
+      return dataPoint;
+    });
+  }
+
+  function saveUpdatedCropData(e) {
+    e.preventDefault();
+    onChange(cropData);
+  }
+
+  return (
+    <div className="stage-details-container">
+      {/* Section header */}
+      <SectionHeader onClick={() => null} tabIndex={0}>
+        Stage Details
+      </SectionHeader>
+
+      <form className="stage-values-form" onSubmit={saveUpdatedCropData}>
```

Comment:
There are at least 5 - 10 components nested/hidden in this GIGANTIC!!! function. 
Please, remember the "React way": https://reactjs.org/docs/thinking-in-react.html

This might not be obvious now, but later when you are working on a project where you have 200 components written like this, and the bug´s start to appear (and they will!), the whole thing will collapse. 
Remember what I said 😇, web applications a simple computer programs, but usually with enormous amount of code. This means that your most important considerations, are **readability and maintainability**. If you write code only to be seen by yourself... then do whatever, but when collaborating, everything you write has be understood be everyone else also, and everyone else is busy doing their own stuff. 

1. Whenever you are returning nested divs, its safe to assume that these should be replaced by components.
2. Indentation and inline logic like map() and Object.keys() make the code harder to reason about. This logic should be hidden away behind a layer of abstractions.
3. Designers and other programmers coming from other languages will have a hard time interpreting code written in an imperative style.  A declarative style, on the other hand, which btw, is also the "thinking in react way", serves to explain to the recipient _what_ the  code does and not _how_ javascript works.